OpenSim/Examples/DataTable/example.cpp

```diff
@@ -0,0 +1,357 @@
+// that supports random-access to the columns and rows through the their
+// index. Columns can be labeled with std::string and accessing the columns
+// through labels is constant-time on average. DataTable_ can also hold key-
+// value pairs as metadata where key is a std::string and value can be of any
+// type.
+
+int main() {
+  constexpr double EPSILON{0.00001};
+
+  // Default construct a DataTable of SimTK::Real[alias for double].
+  OpenSim::DataTable_<SimTK::Real> dt_real{};
+  // Default construct a DataTable of SimTK::Vec3[a tuple of 3 double(s)].
+  OpenSim::DataTable_<SimTK::Vec3> dt_vec3{};
+  // Default construct a DataTable of SimTK::Vec6[a tuple of 6 double(s)].
+  OpenSim::DataTable_<SimTK::Vec6> dt_vec6{};
+
+  std::vector<SimTK::Real> data_real{};
+  std::vector<SimTK::Vec3> data_vec3{};
+  std::vector<SimTK::Vec6> data_vec6{};
+  for(int i = 0; i < 12; ++i) {
+    data_real.push_back(i);
+    data_vec3.push_back(SimTK::Vec3{1, 2, 3} + i);
+    data_vec6.push_back(SimTK::Vec6{1, 2, 3, 4, 5, 6} + i);
+  }
+
+  // Construct a DataTable using iterators. We can arrange the elements
+  // row-wise or column-wise using the argument 'dir'. The argument 'ndir' is
+  // required in either case to let the DataTable know about the shape of the
+  // matrix. The following creates a 4x3 DataTable filling the elements
+  // column-wise.
+  {
+    OpenSim::DataTable_<SimTK::Real> real{data_real.cbegin(),
+                                          data_real.cend(),
+                                          3,
+                                          OpenSim::COLWISE};
+    OpenSim::DataTable_<SimTK::Vec3> vec3{data_vec3.cbegin(),
+                                          data_vec3.cend(),
+                                          3,
+                                          OpenSim::COLWISE};
+    OpenSim::DataTable_<SimTK::Vec6> vec6{data_vec6.cbegin(),
+                                          data_vec6.cend(),
+                                          3,
+                                          OpenSim::COLWISE};
+  }
+
+  // Add multiple rows at once using an iterator. Because the DataTable is empty
+  // , number of columns has to specified when using iterator. When calling this
+  // function on a non-empty DataTable, the arugment ncol is ignored. We have 12
+  // elements in the iterator and following creates a 3x4 DataTable.
+  dt_real.addRows(data_real.cbegin(), data_real.cend(), 4);
+  dt_vec3.addRows(data_vec3.cbegin(), data_vec3.cend(), 4);
+  dt_vec6.addRows(data_vec6.cbegin(), data_vec6.cend(), 4);
+
+  std::vector<SimTK::Real> data1_real(4, 1);
+  std::vector<SimTK::Vec3> data1_vec3(4, {1, 2, 3});
+  std::vector<SimTK::Vec6> data1_vec6(4, {1, 2, 3, 4, 5, 6});
+
+  // Add a row to the DataTable using SimTK::RowVector_. Create SimTK::RowVector
+  // and add it as a row. After this the DataTable will of size 4x4.
+  SimTK::RowVector_<SimTK::Real> row_real{static_cast<int>(data1_real.size()), 
```

Comment:
We should add alternate constructor signatures to Matrix, Vector, RowVector that allow `size_t` to avoid these ugly casts.